SAT/Z3Interfacing.hpp

```diff
@@ -141,19 +141,55 @@ class Z3Interfacing : public PrimitiveProofRecordingSATSolver
 
   struct FuncOrPredId
   {
-    explicit FuncOrPredId(unsigned id, bool isPredicate) : id(id), isPredicate(isPredicate) {}
-    explicit FuncOrPredId(Term* term) : FuncOrPredId(term->functor(), term->isLiteral()) {}
-    static FuncOrPredId function(FuncId id) { return FuncOrPredId ( id, false ); }
-    static FuncOrPredId predicate(PredId id) { return FuncOrPredId ( id, true ); }
+    explicit FuncOrPredId(unsigned id, bool isPredicate, Term *forSorts = nullptr) : id(id), isPredicate(isPredicate), forSorts(forSorts) {}
+    explicit FuncOrPredId(Term* term) :
+      FuncOrPredId(
+        term->functor(),
+        term->isLiteral(),
+        term->numTypeArguments() == 0
+          ? nullptr
+          : term
+      )
+    {}
+    static FuncOrPredId monomorphicFunction(FuncId id) { return FuncOrPredId (id, false); }
+    static FuncOrPredId monomorphicPredicate(PredId id) { return FuncOrPredId (id, true); }
     unsigned id;
     bool isPredicate;
+    /**
+     * polymorphic symbol application: treat e.g. f(<sorts>, ...) as f<sorts>(...) for Z3
+     * in the monomorphic case, nullptr
+     * in the polymorphic case, some term of the form f(<sorts>, ...) which we use only for sort information
+     */
+    Term *forSorts;
 
     friend struct std::hash<FuncOrPredId> ;
     friend bool operator==(FuncOrPredId const& l, FuncOrPredId const& r)
-    { return l.id == r.id && l.isPredicate == r.isPredicate; }
+    {
+      if(l.id != r.id || l.isPredicate != r.isPredicate)
+        return false;
+      if(!l.forSorts)
+        return true;
+
+      // compare sort arguments
+      for(int i = 0; i < l.forSorts->numTypeArguments(); i++)
+        // sorts are perfectly shared
+        if(!l.forSorts->nthArgument(i)->sameContent(r.forSorts->nthArgument(i)))
+          return false;
+
+      return true;
```

Comment:
Why can't we check that by just checking whether `l.forSorts == r.forSorts` could it be that two ids get different `forSorts`s that represent the same type arguments?

Reply:
I believe so! If for instance we are trying to insert `f($int, 1)` and `f($int, 2)` then they will have the same type arguments but a different `forSorts`.

Reply:
Aaah right! I see! All good then! I was just confused for a sec.